# Homework 1: Introduction for MLOps Zoomcamp 2025

You can find the homework at [cohorts/2025/01-intro/homework.md](https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/cohorts/2025/01-intro/homework.md "https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/cohorts/2025/01-intro/homework.md").

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import root_mean_squared_error

In [2]:
nyc_yellow_taxi_jan_2023_parquet_url = 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet'
nyc_yellow_taxi_feb_2023_parquet_url = 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet'

In [3]:
yt_jan23_df = pd.read_parquet(nyc_yellow_taxi_jan_2023_parquet_url)

In [4]:
yt_feb23_df = pd.read_parquet(nyc_yellow_taxi_feb_2023_parquet_url)

In [5]:
yt_jan23_df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0         14.30                   2.5         0.00  
1                    1.0         16.90                   2.5         0.00  
2                    1.0         34.90                   2.5         0.00  
3                    1.0         20.85                   0.0         1.25  
4                    1.0         19.68                   2.5         0.00

In [6]:
yt_feb23_df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2023-02-01 00:32:53   2023-02-01 00:34:34              2.0   
1         2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
2         2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
3         1  2023-02-01 00:29:33   2023-02-01 01:01:38              0.0   
4         2  2023-02-01 00:12:28   2023-02-01 00:25:46              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.30         1.0                  N           142           163   
1           0.00         1.0                  N            71            71   
2           0.00         1.0                  N            71            71   
3          18.80         1.0                  N           132            26   
4           3.22         1.0                  N           161           145   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.4   3.50      0.5         0.0           0.0   
1             4         -3.0  -1.00     -0.5         0.0           0.0   
2             4          3.0   1.00      0.5         0.0           0.0   
3             1         70.9   2.25      0.5         0.0           0.0   
4             1         17.0   1.00      0.5         3.3           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  
0                    1.0          9.40                   2.5         0.00  
1                   -1.0         -5.50                   0.0         0.00  
2                    1.0          5.50                   0.0         0.00  
3                    1.0         74.65                   0.0         1.25  
4                    1.0         25.30                   2.5         0.00

In [7]:
yt_jan23_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

## Solutions to the questions

### Question 1.
We'll use the [same NYC taxi dataset](https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page "https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page"), but instead of "Green Taxi Trip Records", we'll use "Yellow Taxi Trip Records". Download the data for January and February 2023. Read the data for January. How many columns are there?

* 16
* 17
* 18
* **19**

In [8]:
yt_jan23_df.shape

(3066766, 19)

### Answer 1.
There are 19 columns in the dataset.

### Question 2.

Now let's compute the `duration` variable. It should contain the duration of a ride in minutes. What's the standard deviation of the trips duration in January?

* 32.59
* **42.59**
* 52.59
* 62.59

In [9]:
duration_diff_sr = yt_jan23_df.tpep_dropoff_datetime - yt_jan23_df.tpep_pickup_datetime
yt_jan23_df['duration'] = duration_diff_sr.apply(lambda x: x.total_seconds()/60)

In [10]:
print(f'''The standard deviation in minutes of the ride duration is {np.round(np.std(yt_jan23_df['duration']), 2)}.''')

The standard deviation in minutes of the ride duration is 42.59.


### Answer 2.

The standard deviation in minutes of the ride duration is 42.59.

### Question 3. Dropping outliers

Next, we need to check the distribution of the `duration` variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive). What fraction of the records left after you dropped the outliers?

* 90%
* 92%
* 95%
* **98%**

In [11]:
duration_df_without_outliers = yt_jan23_df['duration'][(yt_jan23_df['duration'] >= 1) & (yt_jan23_df['duration'] <= 60)]

In [12]:
duration_df_with_outliers = yt_jan23_df['duration']

In [13]:
np.round(100*duration_df_without_outliers.shape[0]/duration_df_with_outliers.shape[0])

np.float64(98.0)

### Answer 3.

The fraction of data without the outliers is 98%.

### Question 4: One-hot encoding

Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

* Turn the dataframe into a list of dictionaries (remember to re-cast the IDs to strings - otherwise it will label encode them)
* Fit a dictionary vectorizer
* Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

* 2
* 155
* 345
* **515**
* 715

In [14]:
jan_wo_df = yt_jan23_df[(yt_jan23_df['duration'] >= 1) & (yt_jan23_df['duration'] <= 60)]
jan_wo_pu_do_loc_df = jan_wo_df[['PULocationID', 'DOLocationID']].astype(str)

In [15]:
jan_wo_pu_do_loc_dict = jan_wo_pu_do_loc_df.to_dict(orient='records')

In [16]:
dv = DictVectorizer()
jan_wo_transf_arr = dv.fit_transform(jan_wo_pu_do_loc_dict)

In [17]:
print(f'The dimensions of the feature matrix is {jan_wo_transf_arr.shape[1]}.')

The dimensions of the feature matrix is 515.


### Answer 4.

The dimensions of the feature matrix is 515.

### Question 5. Training a model

Now let's use the feature matrix from the previous step to train a model.

* Train a plain linear regression model with default parameters
* Calculate the RMSE of the model on the training data

What's the RMSE on train?

* 3.64
* **7.64**
* 11.64
* 16.64

In [18]:
X_jan = jan_wo_transf_arr.copy()
y_jan = jan_wo_df['duration'].copy()

In [19]:
lr = LinearRegression()
lr.fit(X_jan, y_jan)

LinearRegression()

In [20]:
y_jan_pred = lr.predict(X_jan)
print(f'The RMSE on the January data is: {round(root_mean_squared_error(y_jan, y_jan_pred), 2)}.')

The RMSE on the January data is: 7.65.


### Answer 5.

The RMSE on the January data is: 7.65.

### Question 6. Evaluating the model

Now let's apply this model to the validation dataset (February 2023). What's the RMSE on validation?

* 3.81
* **7.81**
* 11.81
* 16.81

In [21]:
feb_duration_diff_sr = yt_feb23_df.tpep_dropoff_datetime - yt_feb23_df.tpep_pickup_datetime
yt_feb23_df['duration'] = feb_duration_diff_sr.apply(lambda x: x.total_seconds()/60)

In [22]:
feb_wo_df = yt_feb23_df[(yt_feb23_df['duration'] >= 1) & (yt_feb23_df['duration'] <= 60)]
feb_wo_pu_do_loc_df = feb_wo_df[['PULocationID', 'DOLocationID']].astype(str)

In [23]:
feb_wo_pu_do_loc_dict = feb_wo_pu_do_loc_df.to_dict(orient='records')

In [24]:
feb_wo_transf_arr = dv.transform(feb_wo_pu_do_loc_dict)

In [25]:
X_feb = feb_wo_transf_arr.copy()
y_feb = feb_wo_df['duration'].copy()

In [26]:
lr.fit(X_feb, y_feb)

LinearRegression()

In [27]:
y_feb_pred = lr.predict(X_feb)
print(f'The RMSE on the February data is: {round(root_mean_squared_error(y_feb_pred, y_feb), 2)}.')

The RMSE on the February data is: 7.78.


### Answer 6.

The RMSE on the February data is: 7.78.